C#: Stub generation

tamasvajk opened the pull request:
This PR improves stub generation:
* added a script which takes a nuget package as an input
* general stubbing is improved to cover a lot more constructs (`delegate`, required non-`public` members, `protected private` and `protected internal` accessibility, ...)
* added dotnet core, asp.net core stubs, and a test referencing them.

~~[Differences job](https://jenkins.internal.semmle.com/job/Changes/job/CSharp-Differences/1107/)~~
~~[Differences job](https://jenkins.internal.semmle.com/job/Changes/job/CSharp-Differences/1110/)~~
~~[Differences job](https://jenkins.internal.semmle.com/job/Changes/job/CSharp-Differences/1132/)~~
https://jenkins.internal.semmle.com/job/Changes/job/CSharp-Differences/1136/

Follow-up PRs, where new stubs are introduced and used:
- https://github.com/tamasvajk/codeql/pull/8 (newtonsoft.json)
- https://github.com/tamasvajk/codeql/pull/9 (dapper + sqlclient)
- https://github.com/tamasvajk/codeql/pull/10 (nhibernate + missing ctor bugfix)

tamasvajk commented:
@hvitved Could you have a first pass review of this one? I'll add more test cases...

hvitved reviewed (commented):
Some initial comments. Looks great!

tamasvajk commented on csharp/ql/src/Stubs/Stubs.qll:
```diff
@@ -76,29 +83,34 @@ abstract private class GeneratedType extends ValueOrRefType, GeneratedElement {
     then result = ""
-    else
+    else (
+      not this instanceof DelegateType and
```
`delegate` declarations look very different from other type declarations. `[access modifier] delegate [return type] [delegate name]([parameters])`. Return type and parameters are specific to `delegates`, so it seemed more readable to handle them in a dedicated way.

tamasvajk commented on csharp/extractor/Semmle.Extraction.CSharp/Extractor/Extractor.cs:
```diff
@@ -129,6 +134,42 @@ public static ExitCode Run(string[] args)
+
+                foreach (var cs in csProj.Sources)
+                {
+                    if (cs.EndsWith("obj/Debug/.NETCoreApp,Version=v5.0.AssemblyAttributes.cs") ||
```
The generated .csproj files are [now hard coded](https://github.com/github/codeql/pull/5664/files?file-filters%5B%5D=.py#diff-37daa16469133530ed373f2fce5f58d735f83fd4ff8d3dcdc5cab49a66be6113R132) to `<TargetFramework>net5.0</TargetFramework>`, so this filtering will work. But it would be great if we didn't have to update all stub .csproj files and this bit of code when we move to newer version of .net. 

I think the file name is generated by [this code](https://github.com/dotnet/msbuild/blob/main/src/Tasks/Microsoft.Common.CurrentVersion.targets#L3417).

Actually, what do you think, shouldn't we simply exclude all files from the `obj` folders?

tamasvajk commented:
The failing CI job should be fixed by https://github.com/github/codeql/pull/5978.

tamasvajk commented:
The differences job shows false positives in `cs/dereferenced-value-is-always-null`.

tamasvajk commented:
I converted this back to a draft PR. I'll split it up into multiple smaller chunks. This will help in the review, and also, I'll trigger the differences job on smaller changes.

tamasvajk commented on csharp/extractor/Semmle.Extraction.CSharp/Extractor/Extractor.cs:
```diff
@@ -399,9 +440,10 @@ private static IEnumerable<Action> ReadSyntaxTrees(IEnumerable<string> sources,
+                        compilerArguments.CompilationOptions
+                            .WithAssemblyIdentityComparer(DesktopAssemblyIdentityComparer.Default)
+                            .WithStrongNameProvider(new DesktopStrongNameProvider(compilerArguments.KeyFileSearchPaths))
+                            .WithMetadataImportOptions(MetadataImportOptions.All)
```
Extracted this change to a dedicated PR: https://github.com/github/codeql/pull/6000

tamasvajk commented on csharp/extractor/Semmle.Extraction.CSharp/Extractor/Options.cs:
```diff
@@ -68,6 +74,9 @@ public override bool HandleOption(string key, string value)
                 case "framework":
                     Framework = value;
                     return true;
+                case "load-sources-from-project":
```
Maybe `add-sources-from-project` is a better name.

tamasvajk commented:
@hvitved I rebased this PR after splitting out some of the changes to separate PRs. The size of the PR is still big because of the added asp.net core and .net core stubs. I also rewrote the git history a bit in order to allow commit-by-commit review. I hope it got to a stage where it's readable enough so that the review is not a huge pain. Let me know if I should adjust further.

tamasvajk commented on csharp/ql/src/Stubs/make_stubs_nuget.py:
```diff
@@ -0,0 +1,220 @@
+stubsDir = os.path.join(outputDir, stubsDirName)
+os.makedirs(stubsDir)
+
+frameworksDirName = 'frameworks'
```
`_frameworks` would be better to keep it above all other stubs in the `stubs` folder.

tamasvajk replied:
@hvitved I pushed an extra commit with this change.

hvitved reviewed (commented):
LGTM. Two small questions.

tamasvajk commented on csharp/ql/src/Stubs/Stubs.qll:
```diff
@@ -12,51 +12,89 @@
 
 import csharp
+private import semmle.code.csharp.frameworks.System
+private import semmle.code.dotnet.DotNet as DotNet
```
I don't think it's needed.

tamasvajk commented:
@hvitved Tests were failing on this because of the adjusted qualified names of nested types in `main`. I rebased this PR, and adjusted the test expected file. (I also regenerated the stubs.)
